### PR TITLE
mesh: Allow VertexOnlyMeshes embedded in 1D

### PR DESCRIPTION
**Pull request by wence-**
Needs https://gitlab.com/petsc/petsc/-/merge_requests/4233

### CODE REVIEW

**Comment by wence-** [comment]
Will add a test once the petsc changes are in.

**Comment by ReubenHill** [comment]
This is great. What's your use case out of interest?

> Will add a test once the petsc changes are in.

I think the tests are there, they just need to be un-xfailed. 
If so, this should add all the tests needed


```diff
diff --git a/tests/vertexonly/test_interpolation_from_parent.py b/tests/vertexonly/test_interpolation_from_parent.py
index 6fc96de4..b100d176 100644
--- a/tests/vertexonly/test_interpolation_from_parent.py
+++ b/tests/vertexonly/test_interpolation_from_parent.py
@@ -9,7 +9,7 @@ from operator import add

 # NOTE we don't include interval mesh since many of the function spaces
 # here are not defined on it.
-@pytest.fixture(params=[pytest.param("interval", marks=pytest.mark.xfail(reason="swarm not implemented in 1d")),
+@pytest.fixture(params=["interval",
                         "square",
                         "squarequads",
                         pytest.param("extruded", marks=pytest.mark.xfail(reason="extruded meshes not supported")),
diff --git a/tests/vertexonly/test_swarm.py b/tests/vertexonly/test_swarm.py
index 3ee6b617..5f91b47b 100644
--- a/tests/vertexonly/test_swarm.py
+++ b/tests/vertexonly/test_swarm.py
@@ -36,7 +36,7 @@ def cell_midpoints(m):
     return midpoints, local_midpoints


-@pytest.fixture(params=[pytest.param("interval", marks=pytest.mark.xfail(reason="swarm not implemented in 1d")),
+@pytest.fixture(params=["interval",
                         "square",
                         pytest.param("extruded", marks=pytest.mark.xfail(reason="extruded meshes not supported")),
                         "cube",
diff --git a/tests/vertexonly/test_vertex_only_fs.py b/tests/vertexonly/test_vertex_only_fs.py
index 6fc14b3f..ea57392b 100644
--- a/tests/vertexonly/test_vertex_only_fs.py
+++ b/tests/vertexonly/test_vertex_only_fs.py
@@ -6,7 +6,7 @@ from mpi4py import MPI

 # Utility Functions

-@pytest.fixture(params=[pytest.param("interval", marks=pytest.mark.xfail(reason="swarm not implemented in 1d")),
+@pytest.fixture(params=["interval",
                         "square",
                         pytest.param("extruded", marks=pytest.mark.xfail(reason="extruded meshes not supported")),
                         "cube",
diff --git a/tests/vertexonly/test_vertex_only_mesh_generation.py b/tests/vertexonly/test_vertex_only_mesh_generation.py
index 605866f5..483c4220 100644
--- a/tests/vertexonly/test_vertex_only_mesh_generation.py
+++ b/tests/vertexonly/test_vertex_only_mesh_generation.py
@@ -36,7 +36,7 @@ def cell_midpoints(m):
     return midpoints, local_midpoints


-@pytest.fixture(params=[pytest.param("interval", marks=pytest.mark.xfail(reason="swarm not implemented in 1d")),
+@pytest.fixture(params=["interval",
                         "square",
                         pytest.param("extruded", marks=pytest.mark.xfail(reason="extruded meshes not supported")),
                         "cube",
```

**Review by ReubenHill — COMMENTED** [review]
I see no issues with the changes as currently proposed. Am keen to see if all the x-fail tests pass

**Comment by ReubenHill** [comment]
Note that this will resolve #1792

**Comment by wence-** [comment]
> Needs https://gitlab.com/petsc/petsc/-/merge_requests/4233

This is in.

File: firedrake/interpolation.py [hunk 1]
```diff
@@ -285,15 +285,15 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
     if not isinstance(expr, firedrake.Expression):
         kernel = compile_expression_dual_evaluation(expr, to_element,
                                                     domain=source_mesh,
-                                                    parameters=parameters,
-                                                    coffee=False)
+                                                    parameters=parameters)
```
Comment (ReubenHill) [inline firedrake/interpolation.py:288]: You need to make this change in `/firedrake/preconditioners/pmg.py` too

File: firedrake/interpolation.py [hunk 2]
```diff
@@ -460,6 +460,9 @@ def composed_map(map1, map2):
     Requires that `map1.toset == map2.iterset`.
     Only currently implemented for `map1.arity == 1`
     """
+    if map2 is None:
+        # Real function space case
+        return None
```
Comment (ReubenHill) [inline firedrake/interpolation.py:465]: Great so this should fix #2095 ? Can you add a test to `tests/regression/test_interpolate.py`?

File: tests/vertexonly/test_interpolation_from_parent.py [hunk 3]
```diff
@@ -88,7 +96,7 @@ def pseudo_random_coords(size):
 # interpolation from a CG1 VectorFunctionSpace (I think)
 def test_scalar_spatialcoordinate_interpolation(parentmesh, vertexcoords):
     vm = VertexOnlyMesh(parentmesh, vertexcoords)
-    vertexcoords = vm.coordinates.dat.data_ro
+    vertexcoords = vm.coordinates.dat.data_ro.reshape(-1, parentmesh.geometric_dimension())
```
Comment (ReubenHill) [inline tests/vertexonly/test_interpolation_from_parent.py:99]: What's the purpose of this change? Accounting for immersed manifolds?

File: tests/vertexonly/test_interpolation_from_parent.py [hunk 4]
```diff
@@ -132,17 +140,10 @@ def test_vector_function_interpolation(parentmesh, vertexcoords, vfs):
         if parentmesh.ufl_cell().cellname() == "quadrilateral":
             if not (vfs_fam == "RTCE" or vfs_fam == "RTCF"):
                 pytest.skip(f"{vfs_fam} does not support {parentmesh.ufl_cell()} cells")
-            # TODO: Remove this else when fixed
-            else:
-                pytest.skip(f"Some complex merge related problem for {vfs_fam}, get this from loopy: TypeError: unsupported type for persistent hash keying: <class 'complex'>")
```
Comment (ReubenHill) [inline tests/vertexonly/test_interpolation_from_parent.py:137]: Is this actually fixed then? What fixed it?

File: tests/vertexonly/test_interpolation_from_parent.py (hunk 3 above)
Comment (wence-) [reply tests/vertexonly/test_interpolation_from_parent.py:99]: PyOP2 collapses the shape of a dat with shape `(whatever, 1)` into `(whatever,)`, so the broadcasting doesn't work.

File: tests/vertexonly/test_vertex_only_mesh_generation.py [hunk 5]
```diff
@@ -148,7 +148,7 @@ def test_generate_cell_midpoints(parentmesh):
         else:
             cell_num = parentmesh.locate_cell(out_of_mesh_point)  # should return None
         if cell_num is not None:
-            assert all(f.dat.data_ro[cell_num] == vm.coordinates.dat.data_ro[i])
+            assert (f.dat.data_ro[cell_num] == vm.coordinates.dat.data_ro[i]).all()
```
Comment (ReubenHill) [inline tests/vertexonly/test_vertex_only_mesh_generation.py:151]: is this preferred style then?

File: tests/vertexonly/test_interpolation_from_parent.py (hunk 4 above)
Comment (wence-) [reply tests/vertexonly/test_interpolation_from_parent.py:137]: Oh sorry, I forgot I also had changes in finat: https://github.com/FInAT/FInAT/pull/92

**Review by ReubenHill — COMMENTED** [review]
See comments. On the whole looks great, will be keen to see if all tests pass one the change to `pmg.py` has been added.

File: tests/vertexonly/test_vertex_only_mesh_generation.py (hunk 5 above)
Comment (wence-) [reply tests/vertexonly/test_vertex_only_mesh_generation.py:151]: when the shape is 1D then the `all(...)` version doesn't work because you get a single numpy.bool back (which is not iterable), but does implement `.all()`

File: tests/vertexonly/test_interpolation_from_parent.py (hunk 3 above)
Comment (ReubenHill) [reply tests/vertexonly/test_interpolation_from_parent.py:99]: Can you add a comment saying as much to the first occurrence of this in each test file please, noting that it's for the interval case?

File: firedrake/interpolation.py (hunk 2 above)
Comment (ReubenHill) [reply firedrake/interpolation.py:465]: Perhaps the tests which you've re-enables in `tests/vertexonly` are enough

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -44,7 +44,7 @@ jobs:
       - name: Build Firedrake
         run: |
           cd ..
-          ./firedrake/scripts/firedrake-install $COMPLEX --venv-name build --tinyasm --disable-ssh --minimal-petsc --slepc --documentation-dependencies --install thetis --install gusto --install icepack --install irksome --install femlium --no-package-manager|| (cat firedrake-install.log && /bin/false)
+          ./firedrake/scripts/firedrake-install $COMPLEX --venv-name build --package-branch tsfc wence/feature/swarm-updates --tinyasm --disable-ssh --minimal-petsc --slepc --documentation-dependencies --install thetis --install gusto --install icepack --install irksome --install femlium --no-package-manager|| (cat firedrake-install.log && /bin/false)
```
Comment (ReubenHill) [inline .github/workflows/build.yml:47]: needs FInAT branch too

Comment (wence-) [reply .github/workflows/build.yml:47]: Now merged.

File: firedrake/interpolation.py (hunk 1 above)
Comment (wence-) [reply firedrake/interpolation.py:288]: Done.

File: firedrake/interpolation.py (hunk 2 above)
Comment (wence-) [reply firedrake/interpolation.py:465]: I think so, that's how I noticed it.

**Comment by wence-** [comment]
I hope I have now correctly enabled all the tests.

**Comment by ReubenHill** [comment]
Looks like the FInAT fix for real function space doesn't work in complex mode

**Comment by wence-** [comment]
Complex fails were a bug in pyop2 which I have fixed, so re-running tests.

**Comment by wence-** [comment]
And fixed, will merge.
